llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +443,80 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
   return tryToDuplicateLoadUse(LoadUses, NonLoadUses, MRI, TII);
 }
 
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// when they are missing.
+bool findGlobalValues(const MachineInstr &MI,
+                      SmallPtrSet<const Value *, 4> &GVSet,
+                      SmallPtrSet<const MachineInstr *, 4> &VisitedInstrs,
+                      unsigned RecMax, MachineRegisterInfo &MRI) {
+
+  // Exhausted the search.
+  if (RecMax == 0)
+    return false;
+
+  // Loaded values cannot be tracked as Globals, but they can
+  // be part of a final address calculation.
+  if (MI.mayLoad())
+    return true;
```

Comment:
Is this safe? I think we should only add an MMO if we could track every possible address

Reply:
What we can observe is that we take an address of a constant table (I consider this they original pointer itself), then we perform some arithmetic on top of the value to select specific values of this table. All those arithmetic  operations are performed using integers (well, vector of integers). Some values that are used in this address calculation are loaded from the memory (maybe parameters). This looks like:

```
    %122:vec256, %189:ep = VLD_pstm_imm_4x32_pseudo %118, 64 :: (load (<16 x s16>) from %ir.it_exp_in.sroa.0.1)
    %488:mss = COPY %124
    %123:vec512 = VFLOOR_S32_BF16_mFl2FxSrc_W %122, %488, implicit-def dead $srf2iflags, implicit $crf2imask
    %125:vec512 = VSHUFFLE %123, %123, %126
    %127:ewl = COPY %125.sub_256_lo
    %128:vec256 = COPY %127
    %487:mss = COPY %95
    %131:acc1024 = VUPS_S64_D16_mv_ups_w2c %128, %487, implicit-def dead $srups_of, implicit $crsat, implicit $crupssign
    %486:mss = COPY %133
    %132:vec512 = VSRS_S32_S64_mv_x_srs %131, %486, implicit-def dead $srsrs_of, implicit $crsat, implicit $crrnd
    %135:vec512 = VBCST_32 %98 <--- this is the first global (in other bb)
    %136:vec512 = VBCST_32 %101 <---- this is the second global  (in other bb)
    %137:vec512 = VSEL_32 %135, %136, %138 <---- select based on 138 (selection mask) / like a shuffle
    %139:vec512 = VADD_32 %137, %132  <---- this is the point where the loaded value is used.
    %140:ewl = COPY %139.sub_256_lo
    %141:ewl = VLDB_4x16_LO %140
    %144:ewh = VLDB_4x16_HI %140
```

Now comes the interesting part of my assumption: 

* We are accessing a global vector here, the base access pointer is used (we track it).
* If we calculate some displacement within this base address:
 -- If we stay in bounds of this address, we are fine, every access to this same address will be handled correctly by the AA.
 -- If we go out of bounds considering this address, then we calculate somethin in excess, so we can end-up in AA problems. However, if we go out-of-bounds, we can blame the input code itself because access out of bounds is UB anyway for C++ so anything can happen.

Considering the load in question, if it gives to us an unsound value that results in a buggy address, it is UB anyway.

Reply:
This makes me a bit uncomfortable, because we have not idea what we are loading, it could actually be other pointers. So potentially, there are other base addresses we do not see.

In the general case, I guess you are right, this load will probably be from a constant pool not representing pointers. I think that as long as we have a pattern like you described, where each vector lane is 32 bits and has been broadcasted with a pointer that we can see, it is okay to assume that the 4x load will be based off of those pointers. If we can't see the VBCST for each lane however, I think we should bail out.

Reply:
Now this case is considered.